design/general_refactorings.md

```diff
@@ -0,0 +1,381 @@
+
+
+
+### Improve Naming
+
+------
+
+##### Proposal
+
+Review the names of files, functions.. and rename them to more readable names.
+
+##### Current State
+
+Not all the names are clear now. For example with `class Resource` it is hard to understand which resource we mean.
+
+##### Recommendation
+
+Go over all naming in the code and rename when need. If we rename  `class Resource` to `class ConjurResource` developer who is reading the code will understand we are referencing a specific resource, a Conjur resource. The same applies for the names of files in the repository.
+
+### Organize directory structure
+
+------
+
+##### Proposal
+
+Update the new current directory structure so that it is divided by the types of files and their responsibilities. For example, all `Controller` classes will be placed in a single directory.
+
+##### Current State
+
+Currently, the directory structure resembles the following
+
+| Directory Name | Purpose                           |
+| :------------- | :-------------------------------- |
+| host           | files related to host command     |
+| init           | files related to init command     |
+| list           | files related to list command     |
+| login          | files related to login command    |
+| logout         | files related to logout command   |
+| policy         | files related to policy command   |
+| user           | files related to user command     |
+| variable       | files related to variable command |
+
+The following files are without parent directory : api.py, client.py, argparse_wrapper.py, cli.py , client.py, config.py, constants.py, credentials_data.py, credentials_form_file.py, endpoint.py, errors.py, http_wrapper.py, resource.py, ssl_service.py, version.py
+
+##### Recommendation
+
+I recommend that we organize the files by their types and not by the commands they represent.
+
+| Directory Name | Purpose                                                      |
+| :------------- | :----------------------------------------------------------- |
+| wrappers       | wrappers and utils of diffrent python libs :  arg_parse_wrapper, http_wrapper |
+| conjur_api     | files that talk directly with conjur. For Example   `conjur_client.py (conjur.py) , conjur_api.py (api.py), ssl_client.py (ssl_service)` |
+| logic          | all logic classes                                            |
+| controller     | all controller classes                                       |
+| logic          | all logic classes                                            |
+| data_objects   | all data objects classes                                     |
+
+
+
+### Improve Error Handling
+
+------
+
+##### Proposal
+
+This will help to better under the error handling flow. This will allow us to catch our own object and not general exceptions that do not proper helpful information to the end user. All Conjur CLI/SDK-specific exceptions should be documented in the funtions that call them.
```

Comment:
replace the first and second sentence with. "Improve our error handling to catch our own Conjur-CLI/SDK specific objects and not general exceptions that do not proper helpful information to the end user."

In this part you are presenting your proposal not your recommendation yet

Reply:
Where should they be documented?

Reply:
> in the funtions that call them.

functions*

Reply:
Fixed The proposal and added the part of documenting the recommendation